notion-core/src/main/scala/com/ambiata/notion/core/InputOutputLocations.scala

```diff
@@ -0,0 +1,106 @@
+package com.ambiata.notion.core
+
+import com.ambiata.mundane.io.{DirPath, FilePath, FileName}
+import org.apache.hadoop.fs.Path
+
+/**
+ * Type for execution locations.
+ *
+ * An execution can only take place locally or on Hdfs
+ */
+sealed trait ExecutionLocation {
+  /** the underlying location */
+  def location: Location
+
+  /** extract a Path for local and hdfs execution in order to pass it to a map-reduce job */
+  def path: Path
+
+  /** the underlying dirPath */
+  def dirPath: DirPath
+
+  /** the underlying filePath */
+  def filePath: FilePath
+}
+
+object ExecutionLocation {
+  def hdfsExecutionLocation(location: HdfsLocation): ExecutionLocation =
+    HdfsExecutionLocation(location)
+
+  def localExecutionLocation(location: LocalLocation): ExecutionLocation =
+    LocalExecutionLocation(location)
+}
+
+case class HdfsExecutionLocation(location: HdfsLocation) extends ExecutionLocation {
+  def </>(fileName: FileName): HdfsExecutionLocation =
+    copy(HdfsLocation((location.dirPath </> fileName).path))
+
+  def </>(path: FilePath): HdfsExecutionLocation =
+    copy(HdfsLocation((location.dirPath </> path).path))
+
+  def path: Path =
+    new Path(location.path)
+
+  def dirPath: DirPath =
+    location.dirPath
+
+  def filePath: FilePath =
+    location.filePath
+}
+
+case class LocalExecutionLocation(location: LocalLocation) extends ExecutionLocation {
+  def </>(fileName: FileName): LocalExecutionLocation =
+    copy(LocalLocation((location.dirPath </> fileName).path))
+
+  def </>(path: FilePath): LocalExecutionLocation =
+    copy(LocalLocation((location.dirPath </> path).path))
+
+  def path: Path =
+    new Path(location.path)
+
+  def dirPath: DirPath =
+    location.dirPath
+
+  def filePath: FilePath =
+    location.filePath
+}
+
+/**
+ * A Location for input / output files
+ *
+ * Some inputs/outputs need to be copied to a second location before being used. We have 5 cases:
+ *
+ *  - local input/output && local execution => no synchronisation required
+ *  - hdfs  input/output && hdfs execution  => no synchronisation required
+ *  - local input/output && hdfs execution  => synchronisation required
+ *  - s3    input/output && local execution => synchronisation required
+ *  - s3    input/output && hdfs execution  => synchronisation required
+ */
+sealed trait SynchronizedLocation {
```

Comment:
Else where we have implemented these types of things as functions over a fold, rather than abstract methods. Any reason for the divergence here?